src/realm/object-store/sectioned_results.cpp

```diff
@@ -216,40 +216,103 @@ struct SectionedResultsNotificationHandler {
     // When set change notifications will be filtered to only deliver
     // change indices refering to the supplied section key.
     util::Optional<Mixed> m_section_filter;
-    // Should be set when the key is StringData or BinaryData.
-    util::Optional<std::shared_ptr<std::string>> m_section_key_buffer;
 };
 
+template <typename StringType>
+void create_buffered_key(Mixed& key, util::Optional<std::string>& buffer, StringType value)
+{
+    if (value.size() == 0) {
+        key = StringType("", 0);
+    }
+    else {
+        buffer = std::string(value.data(), value.size());
+        key = Mixed(buffer);
+    }
+}
+
+ResultsSection::ResultsSection()
+    : m_parent(nullptr)
+    , m_key(Mixed())
+    , m_key_buffer(util::none)
+{
+}
+
+ResultsSection& ResultsSection::operator=(ResultsSection&& other)
+{
+    std::swap(m_parent, other.m_parent);
+    if (other.m_key_buffer) {
+        std::swap(m_key_buffer, other.m_key_buffer);
+        if (other.m_key.is_type(type_String)) {
+            m_key = StringData(m_key_buffer->data(), m_key_buffer->size());
+        }
+        else if (other.m_key.is_type(type_String)) {
+            m_key = BinaryData(m_key_buffer->data(), m_key_buffer->size());
+        }
+        else {
+            REALM_UNREACHABLE();
+        }
+    }
+    else {
+        std::swap(m_key, other.m_key);
+    }
+    return *this;
+}
+
+ResultsSection::ResultsSection(SectionedResults* parent, Mixed key)
+    : m_parent(parent)
+{
+    // Give the ResultsSection its own copy of the string data
+    // to counter the event that the m_previous_str_buffers, m_current_str_buffers
+    // no longer hold a reference to the data.
+    if (key.is_type(type_String, type_Binary)) {
+        key.is_type(type_String) ? create_buffered_key(m_key, m_key_buffer, key.get_string())
+                                 : create_buffered_key(m_key, m_key_buffer, key.get_binary());
+    }
+    else {
+        m_key = key;
+    }
+}
+
+bool ResultsSection::is_valid() const
+{
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
+    m_parent->calculate_sections_if_required();
+    auto it = m_parent->m_sections.find(m_key);
+    return it != m_parent->m_sections.end() && m_parent->is_valid();
+}
+
 Mixed ResultsSection::operator[](size_t idx) const
 {
+    util::CheckedUniqueLock lock(m_parent->m_mutex);
     m_parent->calculate_sections_if_required();
-    REALM_ASSERT(is_valid());
-    auto& section = m_parent->m_sections[m_key];
-    REALM_ASSERT(idx < section.indices.size());
-    return m_parent->m_results.get_any(section.indices[idx]);
+    auto it = m_parent->m_sections.find(m_key);
+    REALM_ASSERT(it != m_parent->m_sections.end());
+    REALM_ASSERT(m_parent->is_valid() && idx < it->second.indices.size());
```

Comment:
Asserting `is_valid()` in these functions is incorrect. If the user accesses a ResultsSection from the wrong thread, that's a bug in their code and not ours so it should be an exception rather than an assertion.

Reply:
Changed the asserts to throw instead.